pkg/odo/genericclioptions/context.go

```diff
@@ -0,0 +1,204 @@
+		util.CheckError(err, "")
+		app = appFlag
+	} else {
+		var err error
+		if !createAppIfNeeded {
+			// otherwise get the current app (which might not exist)
+			app, err = application.GetCurrent(ns)
+		} else {
+			// if we asked an app to be created if missing, get the existing one or creating one if needed
+			app, err = application.GetCurrentOrGetCreateSetDefault(client)
+		}
+		util.CheckError(err, "unable to get current application")
+	}
+
+	// create the internal context representation based on calculated values
+	internalCxt := internalCxt{
+		Client:      client,
+		Project:     ns,
+		Application: app,
+	}
+
+	// create a context from the internal representation
+	context := &Context{
+		internalCxt: internalCxt,
+	}
+
+	// resolve component
+	var cmp string
+	cmpFlag := FlagValueIfSet(command, util.ComponentFlagName)
+	if len(cmpFlag) == 0 {
+		// retrieve the current component if it exists if we didn't set the component flag
+		var err error
+		cmp, err = component.GetCurrent(app, ns)
+		util.CheckError(err, "could not get current component")
+	} else {
+		// if flag is set, check that the specified component exists
+		context.existsOrExit(cmpFlag)
+		cmp = cmpFlag
+	}
+
+	// once the component is resolved, add it to the context
+	context.cmp = cmp
+
+	return context
+}
+
+// FlagValueIfSet retrieves the value of the specified flag if it is set for the given command
+func FlagValueIfSet(cmd *cobra.Command, flagName string) string {
+	flag, err := cmd.Flags().GetString(flagName)
+
+	// log the error for debugging purposes though an error should only occur if the flag hadn't been added to the command or
+	// if the specified flag name doesn't match a string flag. This usually can be ignored.
+	ignoreButLog(err)
+	return flag
+}
+
+// Context holds contextual information useful to commands such as correctly configured client, target project and application
+// (based on specified flag values) and provides for a way to retrieve a given component given this context
+type Context struct {
+	internalCxt
+}
+
+// internalCxt holds the actual context values and is not exported so that it cannot be instantiated outside of this package.
+// This ensures that Context objects are always created properly via NewContext factory functions.
+type internalCxt struct {
+	Client      *occlient.Client
+	Project     string
+	Application string
+	cmp         string
+}
+
+// Component retrieves the optionally specified component or the current one if it is set. If no component is set, exit with
+// an error
+func (o *Context) Component(optionalComponent ...string) string {
+	return o.ComponentAllowingEmpty(false, optionalComponent...)
+}
+
+// ComponentAllowingEmpty retrieves the optionally specified component or the current one if it is set, allowing empty
+// components (instead of exiting with an error) if so specified
+func (o *Context) ComponentAllowingEmpty(allowEmpty bool, optionalComponent ...string) string {
```

Comment:
Can we prefix Get to this , to indicate its a getter?

Reply:
Sure, the idea here, though, was to more or less match the `Client`, `Application`, `Project` mold :)